transport/src/main/java/io/netty/bootstrap/Bootstrap.java

```diff
@@ -256,21 +254,12 @@ public void run() {
 
     @Override
     @SuppressWarnings("unchecked")
-    void init(Channel channel) throws Exception {
+    void init(Channel channel) {
         ChannelPipeline p = channel.pipeline();
         p.addLast(config.handler());
 
-        final Map<ChannelOption<?>, Object> options = options0();
-        synchronized (options) {
-            setChannelOptions(channel, options, logger);
-        }
-
-        final Map<AttributeKey<?>, Object> attrs = attrs0();
-        synchronized (attrs) {
-            for (Entry<AttributeKey<?>, Object> e: attrs.entrySet()) {
-                channel.attr((AttributeKey<Object>) e.getKey()).set(e.getValue());
-            }
-        }
+        setChannelOptions(channel, options0().entrySet().toArray(newOptionArray(0)), logger);
+        setAttributes(channel, attrs0().entrySet().toArray(newAttrArray(0)));
```

Comment:
Why copy these into arrays first? Why not pass and iterate over map directly as is currently done?

Reply:
to minimise the scope of concurrent modifications. I think it is more correct to copy first (just as we do in ServerBootstrap).

Reply:
It's done on the server side for the child options/attrs since they are persisted to be able to apply later (consistently) to every child channel. `CHM.toArray()` just iterates over the set anyway, and I don't imagine it would ever actually be the case that the bootstrap is modified concurrently with channel creation (would be a race bug if so).

So the extra array alloc seems unnecessary to me but I guess it's probably insignificant.

Reply:
@njhill better safe then sorry ;)